### PR TITLE
Add specify order

### PR DESCRIPTION
**Pull request by under-Peter**
This is a work in progress to add order-specification by placing brackets.
While not much time has been invested in making it go fast, it already offers major advantages over a naive `EinCode`, compare e.g.
```julia
julia> a = rand(20,20);

julia> ein"ij,jk,kl -> il"(a,a,a) ≈ ein"(ij,jk),kl -> il"(a,a,a)
true

julia> @btime ein"ij,jk,kl -> il"($a,$a,$a);
  780.942 μs (12 allocations: 3.69 KiB)

julia> @btime ein"(ij,jk),kl -> il"($a,$a,$a);
  64.714 μs (66 allocations: 9.11 KiB)
```
For smaller arrays this advantage is smaller, see e.g.
```julia
julia> a = rand(10,10);
julia> @btime ein"ij,jk,kl -> il"($a,$a,$a);
  50.445 μs (12 allocations: 1.31 KiB)

julia> @btime ein"(ij,jk),kl -> il"($a,$a,$a);
  29.406 μs (66 allocations: 4.36 KiB)
```

Things worth pondering: 
- Does it make sense to have multiple structs or should `NestedEinsum` and `EinCode` be combined?
- Is there a good _intermediate representation_ that we could:
    - efficiently extract from strings
    - efficiently extract from macros
    - efficiently turn into either `EinCode` or `NestedEinsum`

@GiggleLiu , what do you think?

### CODE REVIEW

**Comment by under-Peter** [comment]
### Custom parser vs. Regex

At the moment I'm using a custom parser for a string with parens. In my opinion this is simpler than the regex-expression but that's a matter of opinion. Here I just want to sketch a speed comparison. 
Consider the minimal functions:
```julia
function parse_regex(string, parens = Any[])
    for m in eachmatch(r"(\((?>[^()]|(?R))*\))", string)
        push!(parens, m)
        parse_regex(chop(m.captures[1],head=1,tail=1), parens)
    end
    return parens
end


function parse_fsm(s::AbstractString, i)
    out, g = Any[], Any[]
    while i <= lastindex(s)
        c, j = s[i], nextind(s,i)
        if c === '('
            j, out2 = parse_fsm(s, j)
            push!(out, out2)
        elseif c === ')' || c === ','
            isempty(g) || push!(out, g)
            if  c === ','
                g = Any[]
            else
                return j, out
            end
        elseif isletter(c) # could allow isnumeric() too
            push!(g,c)
        end
        i = j
    end
    isempty(g) || push!(out, g)
    return i, out
end
```

Note that `parse_regex`  does _less_ than `parse_fsm` - from the regex-matches, we'd have to extract the position, figure out which matches belong where, contain which other matches, correspond to which tensors etc. and collect the substrings into tuples. Short: this is a very loose lower bound for performance.

Timings:

 | string  | t_regex (ns) | t_fsm (ns)|
| ------------- | ------------- | -------|
 | "((ij,jk),kl)"| 717| 586 |
 | "((ij,jk),(kl,lm))"| 1027| 800|
 | "(ij,jk,(kl,lm))"| 729| 696|
 | "(ij,(jk,(kl,lm)))"| 1111| 800|

The custom parser is superior while doing much more (preserving the nested structure, extracting the individual indices etc.).
Thus I don't think it's worth pursuing a regex-based solution any further.

Agree @GiggleLiu ?

File: src/einsequence.jl [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+Base.isempty(ig::IndexGroup) = isempty(ig.inds)
+
+"""
+    NestedEinsum
+describes a (potentially) nested einsum. Important fields:
+- `args`, vector of all inputs, either `IndexGroup` objects corresponding to tensors or `NestedEinsum`
+- `iy`, indices of output
+"""
+mutable struct NestedEinsum
+    args::Vector{Union{NestedEinsum, IndexGroup}}
```
Comment (GiggleLiu) [inline src/einsequence.jl]: Why should it be mutable here?

File: src/einsequence.jl [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        end
+        filliys!(arg)
+    end
+    return neinsum
+end
+
+"""
+    IndexGroup
+Leaf in a contractiontree, contains the indices and the number of the tensor it
+describes, e.g. in "ij,jk -> ik", indices "ik" belong to tensor `1`, so
```
Comment (GiggleLiu) [inline src/einsequence.jl]: If I understand correct, `1` means the first input tensor?

File: src/einsequence.jl [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+Base.push!(neinsum::NestedEinsum, x) = (push!(neinsum.args,x); neinsum)
+
+"""
+apply a NestedEinsum to arguments evaluates the nested einsum
+"""
+function (neinsum::NestedEinsum)(xs...; size_dict = nothing)
+    ixs = Tuple(map(extractixs, neinsum.args))
+    iy = Tuple(neinsum.iy)
+    EinCode(ixs,iy)(map(arg -> extractxs(xs, arg), neinsum.args)...)
```
Comment (GiggleLiu) [inline src/einsequence.jl]: Generating types dynamically is not efficient? is there an estimation of the overhead?

Comment (under-Peter) [reply src/einsequence.jl]: I changed that (not elegantly) by introducing another type. This improves performance quite a lot although there are still problems.

File: src/einsequence.jl (hunk 2 above)
Comment (under-Peter) [reply src/einsequence.jl]: Yes, for the nested einsum, each individual `einsum` that's ultimately evaluated needs to be able to get its input, this is the label of the input-tensor.

File: src/einsequence.jl (hunk 1 above)
Comment (under-Peter) [reply src/einsequence.jl]: That's a leftover from earlier. will be removed

**Comment by GiggleLiu** [comment]
According to my benchmark, the version with type parameter is faster than the one without type parameter by `~300ns` on my laptop.

Maybe let's merge it first and see if we can get more feedbacks. ok?

**Comment by GiggleLiu** [comment]
```julia
julia> @profile for i=1:10000 parse_nested("(ij,jk),kl", ['i', 'l']) end

julia> Profile.print(mincount=10)
258 ./task.jl:259; (::getfield(REPL, Symbol("##26#27")){REPL.REPLBackend})()
 258 ...ild/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:117; macro expansion
  258 ...ild/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:85; eval_user_input(::Any, ::REPL.REPLBackend)
   258 ./boot.jl:328; eval(::Module, ::Any)
258 /home/leo/.julia/packages/Atom/BXRAC/src/repl.jl:148; evalrepl(::Module, ::String)
 258 ./logging.jl:491; with_logger
  258 ./logging.jl:395; with_logstate(::getfield(Atom, Symbol("##168#170")){Module}, ::Base.CoreL...
   258 /home/leo/.julia/packages/Atom/BXRAC/src/repl.jl:157; (::getfield(Atom, Symbol("##168#170")){Module})()
    258 /home/leo/.julia/packages/Atom/BXRAC/src/repl.jl:135; repleval(::Module, ::Expr)
     258 ./boot.jl:328; eval
      258 .../share/julia/stdlib/v1.1/Profile/src/Profile.jl:25; top-level scope
       258 ./none:1; macro expansion
        49  /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:12; parse_nested(::String, ::Array{Char,1})
         30 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:33; parse_parens(::String, ::Int64, ::Int64)
          11 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:25; parse_parens(::String, ::Int64, ::Int64)
        35  /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:14; parse_nested(::String, ::Array{Char,1})
         13 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:75; filliys!(::OMEinsum.NestedEinsum)
          12 ./array.jl:2400; intersect
         22 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:78; filliys!(::OMEinsum.NestedEinsum)
          16 ./array.jl:2400; intersect(::Array{Char,1}, ::Array{Char,1})
           12 ./array.jl:2396; _shrink(::Function, ::Array{Char,1}, ::Tuple{Array{Char,1}})
        172 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:15; parse_nested(::String, ::Array{Char,1})
         18  /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:142; stabilize(::OMEinsum.NestedEinsum)
          11 ./abstractarray.jl:2018; map
           11 ./array.jl:548; collect_similar
         40  /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:144; stabilize(::OMEinsum.NestedEinsum)
         104 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:145; stabilize(::OMEinsum.NestedEinsum)
          98 ./abstractarray.jl:2018; map(::Function, ::Array{Union{IndexGroup, NestedEinsum},1})
           96 ./array.jl:548; collect_similar(::Array{Union{IndexGroup, NestedEinsum},1}, ::Base.Ge...
            94 ./array.jl:619; _collect(::Array{Union{IndexGroup, NestedEinsum},1}, ::Base.Generato...
             94 ./generator.jl:47; iterate(::Base.Generator{Array{Union{IndexGroup, NestedEinsum},1},ge...
              93 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:145; (::getfield(OMEinsum, Symbol("##119#120")))(::OMEinsum.NestedEinsum)
               21 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:142; stabilize(::OMEinsum.NestedEinsum)
                16 ./abstractarray.jl:2018; map
                 16 ./array.jl:548; collect_similar
               45 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:144; stabilize(::OMEinsum.NestedEinsum)
               18 /home/leo/jcode/OMEinsum.jl/src/einsequence.jl:145; stabilize(::OMEinsum.NestedEinsum)
                13 ./abstractarray.jl:2018; map(::Function, ::Array{Union{IndexGroup, NestedEinsum},1})
                 13 ./array.jl:548; collect_similar(::Array{Union{IndexGroup, NestedEinsum},1}, ::Bas...
```

For reference, this is the profile result of the version with type parameters. Most parsing time is spent on `stablize` function. And this parsing time is lowered by 6us if not using type parameters.

**Comment by under-Peter** [comment]
For
```
@btime ein"(ij,jk),kl -> il"($a,$b,$c)
```
where the matrices are 2x2, I get `2μs` without type parameters, and `3μs` with. Not too much of a difference but if it works it should be alright.
I would be ok with merging then and if we figure out how to parse (i.e. combine it with #50 ) we can merge that one too.

**Pull request merged**